src/cmd-koji-upload

```diff
@@ -117,6 +122,104 @@ class Build(_Build):
     """
     Koji implementation of Build.
     """
+    def __init__(self, *args, **kwargs):
+        self._tmpdir = tempfile.mkdtemp(prefix="koji-build")
+        _Build.__init__(self, *args, **kwargs)
+
+    def __del__(self):
+        try:
+            shutil.rmtree(self._tmpdir)
+            print("removed %s", self._tmpdir)
+        except Exception as e:
+            raise Exception("failed to remove temporary directory: %s",
+                            self._tmpdir, e)
+
+    def rename_mutator(self, fname):
+        """
+        If a file needs to be renamed because of a Koji rule, rename
+        the file to a `.raw`. For some types, its technically incorrect.
+
+        :param fname: file name to check if it needs to be renamed
+        :type str
+        :return srt
+        """
+        for ending in RENAME_RAW:
+            if fname.endswith(ending):
+                return "%s.raw", True
+        return fname, None
+
+    def decompress_mutator(self, fname):
+        """
+        Calculate the mutated name and return a file object suitable
+        for reading the file. If the file is supported as is by Koji,
+        the original file is returned.
+
+        :param fname: name of the local file
+        :type str
+        :return str
+        :return file
+        """
+        for x in COMPRESSION_EXT:
+            if not fname.endswith(x):
+                continue
+
+            base_name = os.path.basename(fname)
+            base = os.path.splitext(base_name)[0] # strip off z off x.y.z
+            ftype = (os.path.splitext(base)[1]).replace(".", "")     # get the y in x.y
+            ctype = "%s.%s" % (ftype, x)
+            if ctype not in KOJI_CG_TYPES and ftype in KOJI_CG_TYPES:
+                if x == "gz":
+                    compressor = gzip
+                elif x == "xz":
+                    raise Exception("not supported yet")
+
+                new_path = os.path.join(self._tmpdir, base)
+                with compressor.open(fname) as infile:
+                    new_path = os.path.join(self._tmpdir, base)
+                    log.info("using %s module to mutate %s to %s",
+                             compressor.__name__, base_name, new_path)
+                    with open(new_path, 'wb+') as outfile:
+                        outfile.write(infile.read())
```

Comment:
This is very inefficient as we end up reading a gigantic buffer only to write it again.  Should use `shutil.copyfileobj()`

Reply:
Indeed, it is. Fixed up.